rdkPlugins/OOMCrash/source/OOMCrashPlugin.cpp

```diff
@@ -0,0 +1,247 @@
+ */
+bool OOMCrash::postHalt()
+{
+    if (!mContainerConfig)
+    {
+        AI_LOG_WARN("Container config is null");
+        return false;
+    }
+    
+    bool status = true;
+    if (mUtils->exitStatus != 0)
+        status = checkForOOM();
+    
+    // Remove the crashFile if container exits normally or if no OOM detected
+    if (mUtils->exitStatus == 0 || status == false)
+    {
+	struct stat buffer;
+	std::string path = mContainerConfig->rdk_plugins->oomcrash->data->path;
+	std::string crashFile = path + "/oom_crashed_" + mUtils->getContainerId() + ".txt";
+	
+	if (stat(crashFile.c_str(), &buffer) == 0)
+	{
+		remove(crashFile.c_str());
+		AI_LOG_INFO("%s file removed", crashFile.c_str());
+	}
+    }
+    
+    AI_LOG_INFO("OOMCrash postHalt hook is running for container with hostname %s", mUtils->getContainerId().c_str());
+    return true;
+}
+
+// End hook methods
+
+/**
+ * @brief Should return the names of the plugins this plugin depends on.
+ *
+ * This can be used to determine the order in which the plugins should be
+ * processed when running hooks.
+ *
+ * @return Names of the plugins this plugin depends on.
+ */
+
+std::vector<std::string> OOMCrash::getDependencies() const
+{
+    std::vector<std::string> dependencies;
+    const rt_defs_plugins_oom_crash* pluginConfig = mContainerConfig->rdk_plugins->oomcrash;
+
+    for (size_t i = 0; i < pluginConfig->depends_on_len; i++)
+    {
+        dependencies.push_back(pluginConfig->depends_on[i]);
+    }
+
+    return dependencies;
+}
+
+/**
+ * @brief Read cgroup file.
+ *
+ *  @param[out]  val      gives the number of times that the cgroup limit was exceeded.
+ *
+ * @return true on successfully reading from the file.
+ */
+
+bool OOMCrash::readCgroup(unsigned long *val)
+{
+    std::string path = "/sys/fs/cgroup/memory/" + mUtils->getContainerId() + "/memory.failcnt";
+
+    FILE *fp = fopen(path.c_str(), "r");
+    if (!fp)
+    {
+        if (errno != ENOENT)
+            AI_LOG_ERROR("failed to open '%s' (%d - %s)", path.c_str(), errno, strerror(errno));
+
+        return false;
+    }
+
+    char* line = nullptr;
+    size_t len = 0;
+    ssize_t rd;
+
+    if ((rd = getline(&line, &len, fp)) < 0)
+    {
+        if (line)
+            free(line);
+        fclose(fp);
+        AI_LOG_ERROR("failed to read cgroup file line (%d - %s)", errno, strerror(errno));
+        return false;
+    }
+
+    *val = strtoul(line, nullptr, 0);
+
+    fclose(fp);
+    free(line);
+
+    return true;
+}
+
+/**
+ * @brief Check for Out of Memory by reading cgroup file.
+ *
+ * @return false when OOM not detected.
+ */
+
+bool OOMCrash::checkForOOM()
+{
+    unsigned long failCnt;
+    bool status;
+    if (readCgroup(&failCnt) && (failCnt > 0))
+    {
+        AI_LOG_WARN("memory allocation failure detected in %s container, likely OOM (failcnt = %lu)", mUtils->getContainerId().c_str(), failCnt);
+        status = createFileForOOM(); 
```

Comment:
I think createFileForOOM() should be called in PostHalt() hook, after the checkForOOM() call.